### PR TITLE
Add reader for Trivista .tvf format

### PR DESCRIPTION
**Pull request by pietsjoh**
### Description of the change
Add support for reading the .tvf-format from trivista.

### Progress of the PR
- [x] basic load functionality (data, axes, original_metadata)
- [x] metadata mapping
- [x] provide option to load glued data as stack
- [x] metadata mapping when loading all metadata
- [x] time series
- [x] axis units
- [x] update docstring (if appropriate),
- [x] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/rosettasciio/blob/main/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests
- [x] check on metadata
- [x] clean up commit history
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
from rsciio.trivista import api
api.file_reader("test.tvf")
```

### CODE REVIEW

File: docs/supported_formats/trivista.rst [hunk 1]
```diff
@@ -0,0 +1,19 @@
+.. _tvf-format:
+
+Trivista
+--------
+
+Reader for spectroscopy data saved using Trivista's software.
```
Comment (jlaehne) [inline docs/supported_formats/trivista.rst:6]:
```suggestion
Reader for spectroscopy data saved using the _Vista Control_ software
for _Trivista_ spectrometers from Teledyne Princeton Instruments.
```

File: rsciio/tests/test_trivista.py [hunk 2]
```diff
@@ -0,0 +1,2373 @@
+# -*- coding: utf-8 -*-
+# Copyright 2007-2022 The HyperSpy developers
+#
+# This file is part of HyperSpy.
```
Comment (jlaehne) [inline rsciio/tests/test_trivista.py:4]: I just see it is still the old HyperSpy header, can you use the new RosettaSciIO header from one of the other files in the package in the newly contributed files.

**Comment by jlaehne** [comment]
Note that we are changing the API slightly to expose the `file_reader` directly without referencing `.api` for each plugin, additionally we provide some standard strings for the docstrings and docstrings will be loaded into the user guide.

The documentation is adapted in  #62

Would be great if you could adapt the PR sometime along the way (if anything is unclear, please leave feedback for #62).

File: docs/supported_formats/trivista.rst [hunk 3]
```diff
@@ -0,0 +1,18 @@
+.. _trivista-format:
+
+TriVista
+--------
+
+Reader for spectroscopy data saved using the VistaControl software
+for TriVista spectrometers from Teledyne Princeton Instruments.
+Currently, RosettaSciIO can only read the ``.tvf`` format from TriVista.
```
Comment (jlaehne) [inline docs/supported_formats/trivista.rst:8]:
```suggestion
Currently, RosettaSciIO can only read the XML-based ``.tvf`` format
from TriVista (the binary ``.tvb`` format is not supported).
```

File: docs/supported_formats/trivista.rst [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+However, this format supports spectral maps and contains all relevant metadata.
+
+If `LumiSpy <https://lumispy.org>`_ is installed, ``LumiSpectrum`` will be
```
Comment (jlaehne) [inline docs/supported_formats/trivista.rst:11]: ``Luminescence`` is the ``signal_type``, while ``LumiSpectrum`` the class name.
See https://docs.lumispy.org/en/latest/user_guide/introduction.html
(needs to be corrected in JobinYvon reader as well)

File: rsciio/trivista/__init__.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from ._api import file_reader
+
+
+__all__ = [
+    "file_reader",
+]
+
+
+def __dir__():
+    return sorted(__all__)
```
Comment (jlaehne) [inline rsciio/trivista/__init__.py:10]: Add to `tests/test_import.py`

File: rsciio/trivista/specifications.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+name: Trivista
```
Comment (jlaehne) [inline rsciio/trivista/specifications.yaml:1]:
```suggestion
name: TriVista
```

File: rsciio/trivista/specifications.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+name_aliases: []
+description: 'Reads data from Trivista tvf files'
```
Comment (jlaehne) [inline rsciio/trivista/specifications.yaml:3]:
```suggestion
description: 'Reads data from TriVista tvf files'
```

File: upcoming_changes/27.new.rst [hunk 8]
```diff
@@ -0,0 +1 @@
+Add support for reading the ``.tvf``-format from Trivista.
```
Comment (jlaehne) [inline upcoming_changes/27.new.rst:1]:
```suggestion
Add support for reading the ``.tvf``-format from TriVista.
```

File: rsciio/trivista/_api.py [hunk 9]
```diff
@@ -0,0 +1,744 @@
...
+    use_uniform_signal_axis: bool, default=False
+        Can be specified to choose between non-uniform or uniform signal axes.
+        If `True`, the ``scale`` attribute is calculated from the average delta along the signal axis
+        and a warning is raised in case the delta varies by more than 1%%.
+    filter_original_metadata: bool, default=True
+        Decides whether to process the original_metadata.
+        If `True`, then non-relevant metadata will be excluded.
+        For example, the metadata usually contains information
+        for multiple objectives, even though only one is used.
+        In this case only the metadata from the used objective
```
Comment (jlaehne) [inline rsciio/trivista/_api.py:704]:
```suggestion
        In this case, only the metadata from the used objective
```

File: rsciio/trivista/_api.py [hunk 10]
```diff
@@ -0,0 +1,744 @@
...
+        will be added to original_metadata.
+        This setting only affects the ``original_metadata`` attribute
+        and not the ``metadata`` attribute.
+    glued_data_as_stack: bool, default=False
+        Using the mode `Step & Glue` results in measurements performed
+        at different wavelength ranges with some overlap between them.
+        The file then contains the individual spectra aswell as
```
Comment (jlaehne) [inline rsciio/trivista/_api.py:711]:
```suggestion
        The file then contains the individual spectra as well as
```

File: rsciio/trivista/_api.py [hunk 11]
```diff
@@ -0,0 +1,744 @@
...
+        the "glued" spectrum. The latter is represented as one spectrum,
+        which covers the complete wavelength range. Stitching the datasets
+        together in the overlap region is already done by the setup.
+        If this setting is set to `True`, then the individual datasets will be loaded as a stack.
+        Otherwise only the "glued" spectrum is loaded.
```
Comment (jlaehne) [inline rsciio/trivista/_api.py:716]:
```suggestion
        Otherwise, only the "glued" spectrum is loaded.
```

File: rsciio/trivista/_api.py [hunk 12]
```diff
@@ -0,0 +1,744 @@
...
+
+_logger = logging.getLogger(__name__)
+
+
+def _error_handling_find_location(len_entry, name):
+    if len_entry > 1:
+        _logger.error(  # pragma: no cover
+            f"File contains multiple positions to read {name} from."
+            "                            "
+            "The first location is choosen."
```
Comment (jlaehne) [inline rsciio/trivista/_api.py:39]:
```suggestion
            "The first location is chosen."
```

**Comment by LMSC-NTappy** [comment]
Absolutely!

If you don't mind I will do it this weekend :)

File: rsciio/trivista/_api.py [hunk 13]
```diff
@@ -0,0 +1,745 @@
...
+#
+# RosettaSciIO is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with RosettaSciIO. If not, see <https://www.gnu.org/licenses/#GPL>.
+
+import xml.etree.ElementTree as ET
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:19]: Do we care about XML vulnerabilities? I don't know much about them but maybe adding a note somewhere about this?

https://docs.python.org/3/library/xml.html#xml-vulnerabilities

(Well, just realized that many other readers make use of this module too...)

File: rsciio/trivista/_api.py [hunk 14]
```diff
@@ -0,0 +1,745 @@
...
+        filter_original_metadata: bool
+            if True then 2 seperate dicts are created
+            (original_metadata, unfiltered_original_metadata).
+            Otherwise only the former is created.
+            Filtering original_metadata is done independent
+            of the value of this parameter.
+            This ensures independence of the metadata mapping
+            on this setting.
+        """
+        self.original_metadata = dict()
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:184]: Consider moving declaration in __init__()

File: rsciio/trivista/_api.py [hunk 15]
```diff
@@ -0,0 +1,745 @@
...
+        ## root level metadata
+        self.original_metadata.update(_etree_to_dict(et_root, only_top_lvl=True))
+        et_fileInfoSerialized = ET.fromstring(
+            self.original_metadata["XmlMain"]["FileInfoSerialized"]
+        )
+        fileInfoSerialized = _etree_to_dict(et_fileInfoSerialized, only_top_lvl=False)
+        self.original_metadata["XmlMain"]["FileInfoSerialized"] = fileInfoSerialized
+
+        ## Documents / Document section
+        self.data_head = et_root[1][0]
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:196]: consider declaring in __init__ first

File: rsciio/trivista/_api.py [hunk 16]
```diff
@@ -0,0 +1,745 @@
...
+        infoSerialized_processed = _process_info_serialized(
+            infoSerialized["Info"]["Groups"]["Group"]
+        )
+        self.original_metadata["Document"]["InfoSerialized"] = infoSerialized_processed
+
+        ## Hardware section
+        metadata_head = et_root[0]
+        metadata_hardware = _etree_to_dict(metadata_head, only_top_lvl=False)
+        if not filter_original_metadata:
+            self.unfiltered_original_metadata = deepcopy(self.original_metadata)
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:211]: consider declaring in `__init__` first

File: rsciio/trivista/_api.py [hunk 17]
```diff
@@ -0,0 +1,745 @@
...
+        ## result for one spectrometer:
+        ## Spectrometers
+        ##     - Spectrometer
+        ##         - ...
+        ## result for 2 spectrometers:
+        ## Spectrometers
+        ##     - Spectrometer1
+        ##         - ...
+        ##     - Spectrometer2
+        ##         - ...
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:261]: I don't see the point in having a different naming convention for 1 or multiple spectrometer. It creates confusion compared to having `Spectrometer.Spectrometer1` for only 1 spectro in my opinion (and also starting numbering from 1 and not 0 is a bit odd but after all why not).

File: rsciio/trivista/_api.py [hunk 18]
```diff
@@ -0,0 +1,745 @@
...
+            try:
+                detector["integration_time"] = (
+                    detector["exposure_per_frame"] * detector["frames"]
+                )
+            except TypeError:  # pragma: no cover
+                pass  # pragma: no cover
+        elif detector["processing"]["calc_average"] == "True":
+            detector["integration_time"] = detector["exposure_per_frame"]
+
+        self._total_frames = int(detector_original.get("No_of_Frames"))
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:362]: consider declaring  in `__init__`

File: rsciio/trivista/_api.py [hunk 19]
```diff
@@ -0,0 +1,745 @@
...
+        laser = self._map_laser_md(laser_wavelength)
+        spectrometer = self._map_spectrometer_md(central_wavelength)
+
+        acquisition_instrument = {
+            "Detector": detector,
+            "Laser": laser,
+        }
+        acquisition_instrument.update(spectrometer)
+
+        self.metadata = {
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:452]: consider declaring in `__init__` f

File: rsciio/trivista/_api.py [hunk 20]
```diff
@@ -0,0 +1,745 @@
...
+            self.data[i] = dataset_reshaped
+
+    def read_file(
+        self,
+        use_uniform_signal_axis=False,
+        glued_data_as_stack=False,
+        filter_original_metadata=True,
+    ):
+        """Calls everything to read the file."""
+        self._use_uniform_signal_axis = use_uniform_signal_axis
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:673]: consider declaring in __init__()

File: rsciio/trivista/_api.py [hunk 21]
```diff
@@ -0,0 +1,745 @@
...
+        time = (np.array(time_array, dtype=np.int64) - time_array[0]).ravel() / 1e7
+        return data, time
+
+    def load_glued_data_stack(self):
+        num_datasets_list = self.data_head.findall("Childs")
+        _error_handling_find_location(
+            len(num_datasets_list), "glued datasets"
+        )  # pragma: no cover
+        data_array = []
+        self.signal_axis_list = []
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:630]: Consider moving to `__init__()`

File: rsciio/trivista/_api.py [hunk 22]
```diff
@@ -0,0 +1,745 @@
...
+            data_array.append(data)
+            time_array.append(time)
+        self.data = np.array(data_array)
+        self.time = np.array(time_array)
+
+    def get_data(self, glued_data_as_stack):
+        if glued_data_as_stack and self.num_datasets != 0:
+            self.load_glued_data_stack()
+        else:
+            data, self.time = self._parse_data(self.data_head)
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:645]: Consider moving to `__init__()`

File: rsciio/trivista/_api.py [hunk 23]
```diff
@@ -0,0 +1,745 @@
...
+        data_array = []
+        self.signal_axis_list = []
+        time_array = []
+        for dataset in num_datasets_list[0]:
+            signal_axis = self._get_signal_axis(dataset)
+            self.signal_axis_list.append(signal_axis)
+            data, time = self._parse_data(dataset)
+            data_array.append(data)
+            time_array.append(time)
+        self.data = np.array(data_array)
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:638]: Consider moving to `__init__()`

File: rsciio/trivista/_api.py [hunk 24]
```diff
@@ -0,0 +1,745 @@
...
+        Parameters
+        ----------
+        filter_original_metadata: bool
+            if True then 2 seperate dicts are created
+            (original_metadata, unfiltered_original_metadata).
+            Otherwise only the former is created.
+            Filtering original_metadata is done independent
+            of the value of this parameter.
+            This ensures independence of the metadata mapping
+            on this setting.
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:182]: This is not the behavior that was retained in the end I think. Consider updating

File: rsciio/trivista/_api.py [hunk 25]
```diff
@@ -0,0 +1,745 @@
...
+        if glued_data_as_stack and self.num_datasets != 0:
+            if has_extra_axis:
+                _logger.warning(  # pragma: no cover
+                    "Loading glued data as stack in combination with multiple axis (time, linescan or map)"
+                    "                            "
+                    "is not tested and may lead to false results."
+                    "                            "
+                    "Please use glued_data_as_stack=False for loading the file."
+                )
+            self.axes = []
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:586]: Consider moving to `__init__()`

File: rsciio/trivista/_api.py [hunk 26]
```diff
@@ -0,0 +1,745 @@
...
+        If this setting is set to `True`, then the individual datasets will be loaded
+        as a stack. Otherwise, only the "glued" spectrum is loaded.
+
+    %s
+    """
+    t = TrivistaTVFReader(Path(filename))
+    t.read_file(
+        use_uniform_signal_axis=use_uniform_signal_axis,
+        glued_data_as_stack=glued_data_as_stack,
+        filter_original_metadata=filter_original_metadata,
```
Comment (LMSC-NTappy) [inline rsciio/trivista/_api.py:725]: Here as well, if you define a _use_uniform_signal_axis instance attribute you should put it in the __init__() initialize it there

**Review by LMSC-NTappy — COMMENTED** [review]
Great Job! Always nice to see new file formats being supported.

Looks pretty good to me, except that I don't like all the instance variables that get initialized randomly in the `TrivistaTVFReader` functions.

I suggest you move all those in the __init__() for clarity.

Cheers

Nicolas

File: rsciio/trivista/_api.py (hunk 17 above)
Comment (pietsjoh) [reply rsciio/trivista/_api.py:307]: When only 1 Spectrometer is used, I wanted that the format of ``metadata`` is consistent with LumiSpy (Acquisition_instrument/Spectrometer/...). Using the same format for ``original_metadata`` makes the mapping more convenient.

Comment (LMSC-NTappy) [reply rsciio/trivista/_api.py:307]: Fair enough.

**Comment by pietsjoh** [comment]
Thanks for the Review @LMSC-NTappy.

I agree that declaring attributes in ``__init__`` gives a better overview of the instance variables. While thinking about setting all instance variables to `None` in ``__init__``, I figured that the read_file() method could be moved to ``__init__`` entirely. I also changed some methods such that they rather return values instead of setting instance variables. These two changes make it possible to set all instance variables directly in the constructor. This also emphasizes the importance of the order in which the methods are called and it makes their dependencies more clear.

Let me know what you think of this solution.

**Review by LMSC-NTappy — APPROVED** [review]
> Let me know what you think of this solution.

Definitely better! Moving everything in the constructor is definitely going some steps further than asked but I don't see it as a problem :)

Someone maintaining this code 5-10 years from now will be thankful for that.

**Review by jlaehne — APPROVED** [review]

**Pull request merged**